### ISSUE TITLE
CfnCluster generated CloudFormation template

### ISSUE THREAD

**Issue by yuanzhou** [awslabs/cfncluster#100]
So I created a cluster with cfncluster, when I went to CloudFormation on the web console, I can see the generated cloudformation template JSON as well as parameters used for customization. I saved the template to a JSON file, and thought maybe I can just upload this template to cloudformation to create another cluster stack. I specified exactly the same parameter values after uploaded the template. And the stack was being created. But I rolled back after the compute node was initiated, very close to success. But still failed even though I tried several times.

I thought this should work since cfncluster calls the cloudformation api. Any troubleshooting suggestions? 

Below is the failure point and after:

```
16:10:03 UTC-0400   ROLLBACK_COMPLETE   AWS::CloudFormation::Stack  yuan    
16:10:02 UTC-0400   DELETE_COMPLETE AWS::SQS::Queue SQS 
16:09:32 UTC-0400   DELETE_COMPLETE AWS::DynamoDB::Table    DynamoDBTable   
16:09:23 UTC-0400   DELETE_COMPLETE AWS::EC2::SecurityGroup MasterSecurityGroup 
16:09:21 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::SecurityGroup MasterSecurityGroup 
16:09:18 UTC-0400   DELETE_COMPLETE AWS::EC2::NetworkInterface  MasterENI   
16:09:18 UTC-0400   DELETE_COMPLETE AWS::EC2::Volume    SharedVolume    
16:09:01 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::NetworkInterface  MasterENI   
16:09:01 UTC-0400   DELETE_IN_PROGRESS  AWS::DynamoDB::Table    DynamoDBTable   
16:09:01 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::Volume    SharedVolume    
16:09:01 UTC-0400   DELETE_IN_PROGRESS  AWS::SQS::Queue SQS 
16:08:58 UTC-0400   DELETE_COMPLETE AWS::EC2::Instance  MasterServer    
16:08:00 UTC-0400   DELETE_COMPLETE AWS::EC2::SecurityGroup ComputeSecurityGroup    
16:07:59 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::SecurityGroup ComputeSecurityGroup    
16:07:58 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::Instance  MasterServer    
16:07:56 UTC-0400   DELETE_COMPLETE AWS::SNS::Topic SNS 
16:07:56 UTC-0400   DELETE_COMPLETE AWS::AutoScaling::LaunchConfiguration   ComputeServerLaunchConfig   
16:07:55 UTC-0400   DELETE_IN_PROGRESS  AWS::SNS::Topic SNS 
16:07:55 UTC-0400   DELETE_IN_PROGRESS  AWS::AutoScaling::LaunchConfiguration   ComputeServerLaunchConfig   
16:07:53 UTC-0400   DELETE_COMPLETE AWS::AutoScaling::AutoScalingGroup  ComputeFleet    
16:05:45 UTC-0400   DELETE_COMPLETE AWS::EC2::EIP   MasterEIP   
16:05:44 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::EIP   MasterEIP   
16:05:41 UTC-0400   DELETE_COMPLETE AWS::EC2::EIPAssociation    AssosiateEIP    
16:05:25 UTC-0400   DELETE_COMPLETE AWS::EC2::SecurityGroupIngress  MasterSecurityGroupIngress  
16:05:24 UTC-0400   DELETE_COMPLETE AWS::EC2::SecurityGroupIngress  ComputeSecurityGroupIngress 
16:05:23 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::SecurityGroupIngress  MasterSecurityGroupIngress  
16:05:22 UTC-0400   DELETE_COMPLETE AWS::SQS::QueuePolicy   SQSPolicy   
16:05:22 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::SecurityGroupIngress  ComputeSecurityGroupIngress 
16:05:22 UTC-0400   DELETE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  ComputeFleet    
16:05:22 UTC-0400   DELETE_IN_PROGRESS  AWS::EC2::EIPAssociation    AssosiateEIP    
16:05:22 UTC-0400   DELETE_IN_PROGRESS  AWS::SQS::QueuePolicy   SQSPolicy   
16:05:19 UTC-0400   ROLLBACK_IN_PROGRESS    AWS::CloudFormation::Stack  yuan    The following resource(s) failed to create: [ComputeFleet]. . Rollback requested by user.
16:05:17 UTC-0400   CREATE_FAILED   AWS::AutoScaling::AutoScalingGroup  ComputeFleet    Received FAILURE signal with UniqueId i-ca6ca650
```

And this is the full list of events before the failure point.

```
15:59:08 UTC-0400   CREATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  ComputeFleet    Resource creation Initiated
15:59:07 UTC-0400   CREATE_IN_PROGRESS  AWS::AutoScaling::AutoScalingGroup  ComputeFleet    
15:58:59 UTC-0400   CREATE_COMPLETE AWS::AutoScaling::LaunchConfiguration   ComputeServerLaunchConfig   
15:58:59 UTC-0400   CREATE_IN_PROGRESS  AWS::AutoScaling::LaunchConfiguration   ComputeServerLaunchConfig   Resource creation Initiated
15:58:57 UTC-0400   CREATE_IN_PROGRESS  AWS::AutoScaling::LaunchConfiguration   ComputeServerLaunchConfig   
15:58:49 UTC-0400   CREATE_COMPLETE AWS::EC2::Instance  MasterServer    
15:58:46 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::Instance  MasterServer    Received SUCCESS signal with UniqueId i-2b6ba1b1
15:53:21 UTC-0400   CREATE_COMPLETE AWS::EC2::EIPAssociation    AssosiateEIP    
15:53:08 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::Instance  MasterServer    Resource creation Initiated
15:53:07 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::Instance  MasterServer    
15:53:06 UTC-0400   CREATE_COMPLETE AWS::EC2::SecurityGroupIngress  MasterSecurityGroupIngress  
15:53:05 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::EIPAssociation    AssosiateEIP    Resource creation Initiated
15:53:05 UTC-0400   CREATE_COMPLETE AWS::EC2::SecurityGroupIngress  ComputeSecurityGroupIngress 
15:53:05 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroupIngress  MasterSecurityGroupIngress  Resource creation Initiated
15:53:04 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroupIngress  MasterSecurityGroupIngress  
15:53:04 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::EIPAssociation    AssosiateEIP    
15:53:04 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroupIngress  ComputeSecurityGroupIngress Resource creation Initiated
15:53:03 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroupIngress  ComputeSecurityGroupIngress 
15:52:58 UTC-0400   CREATE_COMPLETE AWS::EC2::NetworkInterface  MasterENI   
15:52:57 UTC-0400   CREATE_COMPLETE AWS::EC2::SecurityGroup ComputeSecurityGroup    
15:52:55 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroup ComputeSecurityGroup    Resource creation Initiated
15:52:45 UTC-0400   CREATE_COMPLETE AWS::DynamoDB::Table    DynamoDBTable   
15:52:41 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::NetworkInterface  MasterENI   Resource creation Initiated
15:52:40 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::NetworkInterface  MasterENI   
15:52:38 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroup ComputeSecurityGroup    
15:52:37 UTC-0400   CREATE_COMPLETE AWS::SQS::QueuePolicy   SQSPolicy   
15:52:37 UTC-0400   CREATE_IN_PROGRESS  AWS::SQS::QueuePolicy   SQSPolicy   Resource creation Initiated
15:52:36 UTC-0400   CREATE_IN_PROGRESS  AWS::SQS::QueuePolicy   SQSPolicy   
15:52:35 UTC-0400   CREATE_COMPLETE AWS::EC2::Volume    SharedVolume    
15:52:33 UTC-0400   CREATE_COMPLETE AWS::EC2::SecurityGroup MasterSecurityGroup 
15:52:31 UTC-0400   CREATE_COMPLETE AWS::SNS::Topic SNS 
15:52:30 UTC-0400   CREATE_COMPLETE AWS::EC2::EIP   MasterEIP   
15:52:30 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroup MasterSecurityGroup Resource creation Initiated
15:52:19 UTC-0400   CREATE_IN_PROGRESS  AWS::SNS::Topic SNS Resource creation Initiated
15:52:19 UTC-0400   CREATE_IN_PROGRESS  AWS::SNS::Topic SNS 
15:52:16 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::Volume    SharedVolume    Resource creation Initiated
15:52:16 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::Volume    SharedVolume    
15:52:14 UTC-0400   CREATE_IN_PROGRESS  AWS::DynamoDB::Table    DynamoDBTable   Resource creation Initiated
15:52:13 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::SecurityGroup MasterSecurityGroup 
15:52:13 UTC-0400   CREATE_IN_PROGRESS  AWS::DynamoDB::Table    DynamoDBTable   
15:52:13 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::EIP   MasterEIP   Resource creation Initiated
15:52:13 UTC-0400   CREATE_COMPLETE AWS::SQS::Queue SQS 
15:52:12 UTC-0400   CREATE_IN_PROGRESS  AWS::EC2::EIP   MasterEIP   
15:52:12 UTC-0400   CREATE_IN_PROGRESS  AWS::SQS::Queue SQS Resource creation Initiated
15:52:11 UTC-0400   CREATE_IN_PROGRESS  AWS::SQS::Queue SQS 
15:52:05 UTC-0400   CREATE_IN_PROGRESS  AWS::CloudFormation::Stack  yuan    User Initiated
```

**Comment by yuanzhou** [comment]
Tested this template again this morning and it failed at the same point when initializing the first compute node.

**Comment by yuanzhou** [comment]
Very interesting. This time I specified the `InitialQueueSize` and `ComputeWaitConditionCount` as 0 during the parameters screen, with the same template JSON being uploaded before this screen. And it created the cluster successfully. 

So I suspect something can be wrong or mis configed in the template during the creating of the first compute node, when uploading the grnerated template directly to CloudFormation.

**Comment by dougalb** [comment]
When launching using the `cfncluster` CLI, the parameters are read from the configuration file and also some are derived. Looking at the `yuan` stack that you launched, the derived parameters are incorrect. Specifically, the `InitialQueueSize` parameter is set to `1`, while the `ComputeWaitConditionCount` parameter is set to `2`. If you look at the template, the **ComputeFleet** _AutoScalingGroup_ has several properties. The `InitialQueueSize` is used to set the __DesiredCapacity__ (https://github.com/awslabs/cfncluster/blob/master/cloudformation/cfncluster.cfn.json#L2081) and then the __ComputeWaitConditionCount__ is used as part of the __CreationPolicy__ (https://github.com/awslabs/cfncluster/blob/master/cloudformation/cfncluster.cfn.json#L2123).

These two values must be the same during a StackCreate call and afterwards may be different depending on the UpdateStack call.

**Comment by yuanzhou** [comment]
@dougalb Thanks for the reply. Tested again this morning using the template directly in CloudFormation and made sure `ComputeWaitConditionCount` and `InitialQueueSize` set to the same value. The cluster was created successfully without using the CfnCluster CLI. The reason I wanted to test this out the template approach is because sometimes we need to create the cluster but can't install CfnCluster CLI, for example during an online demo.

I've also noticed that the `CLITemplate` won't be used when uploading the template file directly in CloudFormation.
